lib/Frontend/CompilerInvocation.cpp

```diff
@@ -1701,7 +1701,25 @@ static void ParseLangArgs(LangOptions &Opts, ArgList &Args, InputKind IK,
   Opts.PascalStrings = Args.hasArg(OPT_fpascal_strings);
   Opts.VtorDispMode = getLastArgIntValue(Args, OPT_vtordisp_mode_EQ, 1, Diags);
   Opts.Borland = Args.hasArg(OPT_fborland_extensions);
-  Opts.CheckedC = Args.hasArg(OPT_fcheckedc_extension);
+  if (Args.hasArg(OPT_fcheckedc_extension)) {
+    const char *disallowed = nullptr;
+    if (Opts.CUDA)
+      disallowed = "CUDA";
+    else if (Opts.OpenCL)
+      disallowed = "OpenCL";
+    else if (Opts.CPlusPlus)
+      disallowed = "C++";
+    else if (Opts.ObjC1 || Opts.ObjC2)
+      disallowed = "Objective C";
+    if (disallowed) {
+      Diags.Report(diag::err_drv_argument_not_allowed_with) <<
+        "-fcheckedc-extension" << disallowed;
+    } else if (!(Opts.C99 || Opts.C11)) {
+      Diags.Report(diag::err_drv_argument_only_allowed_with) <<
+        "-fcheckedc-extension" << "C";
```

Comment:
Should we be more precise and call out that only C99 and C11 are supported?  I don't remember if clang supports ANSI C under a separate option...